File: server/routers/websocket.py
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Query

from infra.errors import Err
from infra.log import logger
import services.state as state

router = APIRouter()


@router.websocket("/bridge/bot")
async def ws_bot(
    ws: WebSocket,
    token: str = Query(default=""),
):
    bot_token = token or (ws.headers.get("x-astron-bot-token", ""))
    if not await state.token_manager.validate(bot_token):
        await ws.accept()
        await ws.close(code=Err.WS_INVALID_TOKEN.status, reason=Err.WS_INVALID_TOKEN.message)
        logger.warning("Bot connection rejected: invalid token {}...", bot_token[:10])
        return

    if state.bridge.is_draining():
        await ws.accept()
        await ws.close(code=Err.WS_SERVER_RESTART.status, reason=Err.WS_SERVER_RESTART.message)
        logger.info("Bot connection rejected: worker draining token={}...", bot_token[:10])
        return

    await ws.accept()

    client = ws.client
    client_addr = f"{client.host}:{client.port}" if client else "unknown"

    await state.bridge.register_bot(bot_token, ws)

    logger.info("Bot connected: {}... from={}", bot_token[:10], client_addr)
    state.bridge.notify_bot_connected(bot_token)
    try:
        while True:
            raw = await ws.receive_text()
            await state.bridge.handle_bot_message(bot_token, raw)
    except WebSocketDisconnect as exc:
        logger.info("Bot disconnected: {}... from={} code={} reason={}", bot_token[:10], client_addr, exc.code, exc.reason)
    except Exception:
        logger.exception("Bot connection error: {}...", bot_token[:10])
    finally:
        await state.bridge.unregister_bot(bot_token, ws)

File: backend/internal/router/websocket.go
package router

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
	"github.com/rs/zerolog/log"

	"github.com/hygao1024/astron-claw/backend/internal/model"
	"github.com/hygao1024/astron-claw/backend/internal/service"
)

var wsUpgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

func (app *App) wsBot(c *gin.Context) {
	botToken := c.Query("token")
	if botToken == "" {
		botToken = c.GetHeader("x-astron-bot-token")
	}

	if !app.TokenMgr.Validate(c.Request.Context(), botToken) {
		// Accept then close with error code
		conn, err := wsUpgrader.Upgrade(c.Writer, c.Request, nil)
		if err != nil {
			log.Error().Err(err).Msg("WS upgrade failed for invalid token")
			return
		}
		msg := websocket.FormatCloseMessage(model.ErrWSInvalidToken.Code, model.ErrWSInvalidToken.Message)
		_ = conn.WriteMessage(websocket.CloseMessage, msg)
		conn.Close()
		tp := botToken
		if len(tp) > 10 {
			tp = tp[:10]
		}
		log.Warn().Str("token", tp+"...").Msg("Bot connection rejected: invalid token")
		return
	}

	conn, err := wsUpgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		log.Error().Err(err).Msg("WS upgrade failed")
		return
	}

	clientAddr := c.ClientIP()
	botConn := &service.BotConn{
		Conn:  conn,
		Token: botToken,
	}

	ctx := c.Request.Context()
	app.Bridge.RegisterBot(ctx, botToken, botConn)

	tp := botToken[:10]
	log.Info().Str("token", tp+"...").Str("from", clientAddr).Msg("Bot connected")
	app.Bridge.NotifyBotConnected(botToken)

	defer func() {
		app.Bridge.UnregisterBot(ctx, botToken, botConn)
	}()

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				log.Info().Str("token", tp+"...").Str("from", clientAddr).Msg("Bot disconnected normally")
			} else if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				log.Info().Str("token", tp+"...").Str("from", clientAddr).Err(err).Msg("Bot disconnected unexpectedly")
			} else {
				log.Error().Err(err).Str("token", tp+"...").Msg("Bot connection error")
			}
			return
		}

		app.Bridge.HandleBotMessage(ctx, botToken, string(message))
	}
}
